examples/server/tests/features/rerank.feature

```diff
@@ -0,0 +1,43 @@
+@llama.cpp
+@rerank
+Feature: llama.cpp server
+
+  Background: Server startup
+    Given a server listening on localhost:8080
+    And   a model url https://huggingface.co/ggml-org/models/resolve/main/jina-reranker-v1-tiny-en/ggml-model-f16.gguf
+    And   a model file jina-reranker-v1-tiny-en.gguf
+    And   a model alias jina-reranker-v1-tiny-en
+    And   42 as server seed
+    And   2 slots
+    And   512 as batch size
+    And   512 as ubatch size
+    And   512 KV cache size
+    And   embeddings extraction
+    Then  the server is starting
+    Then  the server is healthy
+
+  Scenario: Rerank
+    Given a rerank query:
+      """
+      Machine learning is
+      """
+    And   a rerank document:
+      """
+      A machine is a physical system that uses power to apply forces and control movement to perform an action. The term is commonly applied to artificial devices, such as those employing engines or motors, but also to natural biological macromolecules, such as molecular machines.
+      """
+    And   a rerank document:
+      """
+      Learning is the process of acquiring new understanding, knowledge, behaviors, skills, values, attitudes, and preferences. The ability to learn is possessed by humans, non-human animals, and some machines; there is also evidence for some kind of learning in certain plants.
+      """
+    And   a rerank document:
+      """
+      Machine learning is a field of study in artificial intelligence concerned with the development and study of statistical algorithms that can learn from data and generalize to unseen data, and thus perform tasks without explicit instructions.
+      """
+    And   a rerank document:
+      """
+      Paris, capitale de la France, est une grande ville européenne et un centre mondial de l'art, de la mode, de la gastronomie et de la culture. Son paysage urbain du XIXe siècle est traversé par de larges boulevards et la Seine.
+      """
+    When  reranking request
+    Then  reranking results are returned
+    # TODO: this result make no sense, probably need a better model?
+    Then  reranking highest score is index 3 and lowest score is index 0
```

Comment:
I didn't follow the discussion earlier in the PR, but the test does confirm that the score is quite weird. Probably due to the model size.

Reply:
Does the test add `--pooling rank` when starting the server?

When I run the query manually, it outputs meaningful results:

```bash
curl http://127.0.0.1:8012/v1/rerank \
	-H "Content-Type: application/json" \
	-d '{
      "model": "some-model",
      "query": "Machine learning is",
      "documents": [
        "A machine is a physical system that uses power to apply forces and control movement to perform an action. The term is commonly applied to artificial devices, such as those employing engines or motors, but also to natural biological macromolecules, such as molecular machines.",
        "Learning is the process of acquiring new understanding, knowledge, behaviors, skills, values, attitudes, and preferences. The ability to learn is possessed by humans, non-human animals, and some machines; there is also evidence for some kind of learning in certain plants.",
        "Machine learning is a field of study in artificial intelligence concerned with the development and study of statistical algorithms that can learn from data and generalize to unseen data, and thus perform tasks without explicit instructions.",
        "Paris, capitale de la France, est une grande ville européenne et un centre mondial de lart, de la mode, de la gastronomie et de la culture. Son paysage urbain du XIXe siècle est traversé par de larges boulevards et la Seine."
      ]
    }' | jq
```

```json
{
  "model": "some-model",
  "object": "list",
  "usage": {
    "prompt_tokens": 0,
    "total_tokens": 0
  },
  "results": [
    {
      "index": 0,
      "relevance_score": 0.010812487453222275
    },
    {
      "index": 1,
      "relevance_score": 0.018572865054011345
    },
    {
      "index": 2,
      "relevance_score": 0.030310088768601418
    },
    {
      "index": 3,
      "relevance_score": -0.008471859619021416
    }
  ]
}
```

Reply:
In this case, does it make sense to add an argument `--reranking`?

```cpp
    add_opt(llama_arg(
        {"--reranking"},
        format("enable reranking endpoint on server (default: %s)", params.reranking ? "enabled" : "disabled"),
        [](gpt_params & params) {
            params.reranking = true;
        }
    ).set_examples({LLAMA_EXAMPLE_SERVER}).set_env("LLAMA_ARG_RERANKING"));
```

Then throw an error on `/reranking` endpoint if it's not set:

```cpp
        if (ctx_server.params.reranking) {
            res_error(res, format_error_response("This server does not support reranking. Start it with `--reranking`", ERROR_TYPE_NOT_SUPPORTED));
            return;
        }
```

Reply:
I'm implementing this on my side, will push a commit to this PR

Reply:
Nice - thanks for the help!